### PR TITLE
python-gitlab Issue #63 - implement pagination for list()

### PR DESCRIPTION
**Pull request by jantman**
This implements iteration/pagination of list() results for #63

### CODE REVIEW

File: gitlab/__init__.py [hunk 1]
```diff
@@ -26,6 +26,9 @@
 import requests
 import six
 
+import logging
+logger = logging.getLogger(__name__)
```
Comment (gpocentek) [inline gitlab/__init__.py]: Could you remove the logging code for this change? I'm not against adding some logging but I'd rather do it in another patch.

File: gitlab/__init__.py [hunk 2]
```diff
@@ -189,6 +192,8 @@ def set_url(self, url):
         self._url = '%s/api/v3' % url
 
     def _construct_url(self, id_, obj, parameters):
+        if 'next_url' in parameters:
```
Comment (gpocentek) [inline gitlab/__init__.py]: Could you handle prev, first and last as well?

File: gitlab/__init__.py [hunk 3]
```diff
@@ -342,8 +347,14 @@ def list(self, obj_class, **kwargs):
...
 
-            return [cls(self, item, **cls_kwargs) for item in r.json()
-                    if item is not None]
+            results = [cls(self, item, **cls_kwargs) for item in r.json()
+                       if item is not None]
+            if 'next' in r.links and 'url' in r.links['next']:
+                args = kwargs.copy()
+                args['next_url'] = r.links['next']['url']
+                logger.debug("Iterating results 'next' link: %s", args['next_url'])
+                results.extend(self.list(obj_class, **args))
```
Comment (gpocentek) [inline gitlab/__init__.py]: This doesn't seem to give a choice to the user, this will always return the full list and do multiple API calls. If a user has hundreds of items this is a problem. It should be optional.

Comment (dserodio) [inline gitlab/__init__.py]: IMHO, the perfect implementation would use Python generators instead of a list, and only fetch the next page when needed.

Comment (gpocentek) [inline gitlab/__init__.py]: Sounds like a good plan. I'm a bit short on time to work on this these days but I'd be happy to merge this solution. Thanks!

File: gitlab/__init__.py (hunk 1 above)
Comment (jantman) [inline gitlab/__init__.py]: Done.

File: gitlab/__init__.py (hunk 2 above)
Comment (jantman) [inline gitlab/__init__.py]: @gpocentek I'm not sure I understand the purpose of that, given how this function is called. Even if I don't use the generator idea, from @dserodio, it's your project but I'd rather not add in additional unused code paths...

File: gitlab/__init__.py (hunk 3 above)
Comment (jantman) [inline gitlab/__init__.py]: I'm not sure I see this as "a problem" - the whole reason why I opened this PR is because I have about 200 users, and will soon have about 300 repositories. I've seen some latency when fetching all that, but not much.

I like @dserodio's idea of using a generator for this. I'm going to have a try at implementing it, though I haven't done much with generators before, and perhaps worse, I've never really worked with ``unittest`` or ``httmock``.

**Comment by jantman** [comment]
I've removed the logging code as requested, and added unit tests for the _current_ implementation of just the ``next`` link. However, I think I've hit a bit of a roadblock...

I have nothing against giving a choice to the user, but I'd much rather prefer the iteration to be the default, with disabling it an option. I guess this is personal preference, but I find it very annoying when an API client presents a method like ``Users()`` that only retrieves partial results without making that extremely clear.

I looked into turning this into a generator, but as far as I can tell, I'd need to drastically change how arguments and return values are passed around to get away from the current recursive pattern.

File: gitlab/__init__.py (hunk 2 above)
Comment (gpocentek) [inline gitlab/__init__.py]: My comment was just plain stupid. Sorry.

**Comment by gpocentek** [comment]
Thanks Jason.

I've merged your patch as is but I'll probably restore the default behavior to avoid surprises on updates (people will not expect to suddenly retrieve all the items).

**Comment by jantman** [comment]
Ok, thanks! I'll keep an eye out for it, and update my code to use the pagination once it's released.
